src/drivers/sifive,spi0.c

```diff
@@ -122,43 +140,81 @@ int __metal_driver_sifive_spi0_transfer(struct metal_spi *gspi,
+    METAL_SPI_REGW(METAL_SPI_REG_CSMODE) |= METAL_SPI_CSMODE_HOLD;
 
-        /* Transfer byte */
-        if(tx_buf) {
-            METAL_SPI_REGW(METAL_SPI_REG_TXDATA) = tx_buf[i];
-        } else {
-            METAL_SPI_REGW(METAL_SPI_REG_TXDATA) = 0;
-        }
+    /* Increase the maximum delay between two consecutive frames, in units of clock cycles 
+     * The purpose is to prevent the spi controller from deasserting the cs pin in between
+     * consecutive frames. By default 0 frame interval is allowed between frames. The delay
+     * between each iteration in the for loop is more than one clock cycle. 0xA means the
+     * maximum allowed inter-frame interval is 10 clock cycles.*/
+    /** METAL_SPI_REGW(METAL_SPI_REG_DELAY1) &= ~(0xFF << METAL_SPI_INTERVAL_SHIFT); */
+    /** METAL_SPI_REGW(METAL_SPI_REG_DELAY1) |= (0xF << METAL_SPI_INTERVAL_SHIFT); */
 
-        if(i == (len - 1)) {
-            /* On the last byte, set CSMODE to auto so that the chip select
-             * transitions back to high */
-            METAL_SPI_REGW(METAL_SPI_REG_CSMODE) = METAL_SPI_CSMODE_AUTO;
-        }
+    /* Master send bytes to the slave */
 
-        /* Wait for RXFIFO to not be empty */
-        while((rxdata = METAL_SPI_REGW(METAL_SPI_REG_RXDATA)) & METAL_SPI_RXDATA_EMPTY) ;
+    /* temp_buffer is declared because I cannot modify the value TXDATA register twice by calling &=~ and then |=
+     * instead the final value is stored in temp_buffer, and TXDATA register's value 
+     * will be set to this temporary value */
+    unsigned long temp_buffer;
 
-        if(rx_buf) {
-            rx_buf[i] = (char) (rxdata & 0xFF);
+    for(int i = 0; i < len; i++) {
+        /* Wait for TXFIFO to not be full */
+        while ((temp_buffer = METAL_SPI_REGW(METAL_SPI_REG_TXDATA)) & METAL_SPI_TXDATA_FULL);
+    
+        /* Transfer byte by modifying the least significant byte in the TXDATA register */
+        temp_buffer &= ~(METAL_SPI_TXRXDATA_MASK);
+        temp_buffer |= tx_buf[i];
+        METAL_SPI_REGW(METAL_SPI_REG_TXDATA) = temp_buffer;
+    }
+    
+    /* Reset the maximum interframe delay to 0 clock cycle */
+    /** METAL_SPI_REGW(METAL_SPI_REG_DELAY1) &= ~(0xFF << METAL_SPI_INTERVAL_SHIFT); */
+
+    /* Master receive bytes from the slave */
+    unsigned long rxdata;
+    
+    /* Declare time_t variables to break out of infinite while loop */
+    time_t endwait; 
+
+    for (int i = 0; i < len; i++) {
+        /* Wait for RXFIFO to not be empty, but break the nested loops if timeout, this timeout method 
+         * needs refining, preferably taking into account the device specs */
+        endwait = time(NULL) + METAL_SPI_RXDATA_TIMEOUT;
+        while((rxdata = METAL_SPI_REGW(METAL_SPI_REG_RXDATA)) & METAL_SPI_RXDATA_EMPTY) {
+            if (time(NULL) > endwait) {
+                /* if timeout, immediately stop receiving more bytes */
+                goto end_receive;
```

Comment:
If the receive times out, the method should report an error with a nonzero return code.

@bsousi5, is this how we want to do timeouts in Metal? I just want to make sure this is the best way.

Reply:
I'll change the return code to 1 instead.

Reply:
In the case of a SPI master (which this is) the master supplies the clock.

Think about how "bit-bang-gpio" SPI works, there is no such thing as a timeout.

The hardware SPI module just makes this faster, if there is a timeout - then you have a hardware (chip) problem.

In contrast, for SPI SLAVE - you might consider a very slow clock, or lack of a clock as a timeout but - the protocol does not define a timeout situation.